xde/src/mac/sys.rs

```diff
@@ -177,6 +185,148 @@ unsafe extern "C" {
     ) -> c_int;
     pub fn mac_perim_exit(mph: mac_perim_handle);
     pub fn mac_perim_held(mh: mac_handle) -> boolean_t;
+
+    pub fn mac_hw_emul(
+        mp_chain: *mut *mut mblk_t,
+        otail: *mut *mut mblk_t,
+        ocount: *mut c_uint,
+        mac_emul: u32,
+    );
+
+    pub fn mac_capab_get(
+        mh: *mut mac_handle,
+        capab: mac_capab_t,
+        data: *mut c_void,
+    ) -> boolean_t;
+}
+
+// ======================================================================
+// uts/common/sys/mac_provider.h
+// ======================================================================
+
+#[repr(C)]
+#[derive(Clone, Copy, Default, Debug)]
+pub struct cso_tunnel_t {
+    pub ct_flags: TunnelCsoFlags,
+    pub ct_encap_max: u32,
+    pub ct_types: TunnelType,
+}
+
+#[derive(Clone, Copy, Default, Debug)]
+pub struct mac_capab_cso_t {
+    pub cso_flags: ChecksumOffloadCapabs,
+    pub cso_tunnel: cso_tunnel_t,
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Default, Debug)]
+pub struct lso_basic_tcp_ipv4_t {
+    pub lso_max: t_uscalar_t,
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Default, Debug)]
+pub struct lso_basic_tcp_ipv6_t {
+    pub lso_max: t_uscalar_t,
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Default, Debug)]
+pub struct lso_tunnel_tcp_t {
+    pub tun_pay_max: u32,
+    pub tun_encap_max: u32,
+    pub tun_flags: TunnelTcpLsoFlags,
+    pub tun_types: TunnelType,
+    pub tun_pad: [u32; 2],
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Default, Debug)]
+pub struct mac_capab_lso_t {
+    pub lso_flags: TcpLsoFlags,
+    pub lso_basic_tcp_ipv4: lso_basic_tcp_ipv4_t,
+    pub lso_basic_tcp_ipv6: lso_basic_tcp_ipv6_t,
+
+    pub lso_tunnel_tcp: lso_tunnel_tcp_t,
+}
+
+bitflags::bitflags! {
+/// Classes of TCP segmentation offload supported by a MAC provider.
+#[derive(Clone, Copy, Debug, Default)]
+pub struct TcpLsoFlags: u32 {
+    /// The device supports TCP LSO over IPv4.
+    const BASIC_IPV4 = 1 << 0;
```

Comment:
breadcrumb to `LSO_TX_...` constants in illumos code would be helpful, same for `TunnelTcpLsoFlags` below.